### PR TITLE
Make a last-ditch effort to plan the arrangements necessary for a delta join during LIR planning

### LINKED ISSUE #9921: crash in delta join rendering
### What version of Materialize are you using?

origin/main

### How did you install Materialize?

Built from source

### What was the issue?

running the following code in SLT causes a crash:

```
mode cockroach

statement ok
CREATE TABLE t1 (f1 INTEGER, f2 INTEGER);

statement ok
CREATE TABLE t2  (f1 INTEGER, f2 INTEGER);

statement ok
SELECT 1 FROM t1  RIGHT  JOIN (  SELECT 1 FROM t2  )  ON  NULL  WHERE f2  IN (  9  ,  0  )  OR  0  <  ALL (  SELECT agg1  NULL FROM (  SELECT MAX(  3  ) agg1  FROM (  SELECT  FROM t1  )  GROUP  BY  TRUE  )  )  GROUP  BY  TRUE;
```

### Relevant log output

```source
==> btv.slt                                                                       
thread 'timely:work-0' panicked at 'Arrangement alarmingly absent!: 1, []', src/dataflow/src/render/join/delta_join.rs:75:37
stack backtrace:                                                                  
   0: rust_begin_unwind                                                                                                                                              
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt                                                                                                                                     
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14                                                                 
   2: dataflow::render::join::delta_join::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_delta_join::{{closure}}::{{closure}}::{{closure}}
             at ./src/dataflow/src/render/join/delta_join.rs:75:37                                                                                                   
   3: core::option::Option<T>::unwrap_or_else                                                                                                                        
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/option.rs:787:21
   4: dataflow::render::join::delta_join::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_delta_join::{{closure}}::{{closure}}
             at ./src/dataflow/src/render/join/delta_join.rs:72:35
   5: alloc::collections::btree::map::entry::Entry<K,V>::or_insert_with           
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/collections/btree/map/entry.rs:151:43
   6: dataflow::render::join::delta_join::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_delta_join::{{closure}}
             at ./src/dataflow/src/render/join/delta_join.rs:69:21

(snip)
```

### PR DESCRIPTION
**Pull request by umanwizard**
#9926 mitigated the specific crash described in #9921. However, it may still be the case that bugs in planning or optimization cause us to plan Delta Joins without having access to the expected set of arrangements.

In this case, we should complain loudly in the hopes that we can find out about this when it happens and fix the bugs, but at least we won't crash or produce wrong results.

### Motivation

  * This PR fixes a recognized bug: Fixes #9921 

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. @philip-stoev -- @wangandi 's change in #9926 already mitigated #9921, making it no longer reproduce, so I'm not sure what we can do to test this, if anything.

I've manually verified that this fixes the crash by running it on top of the commit just before #9926. 

- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/dataflow-types/src/plan/mod.rs [hunk 1]
```diff
@@ -397,6 +405,53 @@ impl Plan {
                     // Other plans are errors, and should be reported as such.
                     _ => return Err(()),
                 };
+                let mut required_arrangements = vec![HashSet::new(); inputs.len()];
+                // Delta joins should only be planned if a particular set of arrangements exist.
```
Comment (umanwizard) [inline src/dataflow-types/src/plan/mod.rs:409]:
```suggestion
                // Delta joins should only be planned if a particular set of arrangements exists.
```

File: src/dataflow-types/src/plan/mod.rs [hunk 2]
```diff
@@ -397,6 +405,53 @@ impl Plan {
...
+                // and complain with an error message.
+                if let JoinPlan::Delta(DeltaJoinPlan { path_plans }) = &plan {
+                    for DeltaPathPlan { stage_plans, .. } in path_plans {
+                        for DeltaStagePlan {
+                            lookup_relation,
+                            lookup_key,
+                            ..
+                        } in stage_plans
+                        {
+                            required_arrangements[*lookup_relation].insert(lookup_key.clone());
```
Comment (frankmcsherry) [inline src/dataflow-types/src/plan/mod.rs:421]: Maybe performance golfing, but I bet we could skip the clone if we wanted (`difference` might need to be manually written, but .. also probably skips some more clones).

File: src/dataflow-types/src/plan/mod.rs [hunk 3]
```diff
@@ -397,6 +405,53 @@ impl Plan {
...
+                {
+                    let missing: Vec<_> = required_arrangements
+                        .difference(arrangements)
+                        .cloned()
+                        .collect();
+                    if !missing.is_empty() {
+                        log::error!("Arrangements depended on by delta join alarmingly absent: {:?}
+This is not expected to cause incorrect results, but could indicate a performance issue in Materialize.", missing);
+                        let base_plan =
+                            std::mem::replace(plan, Plan::Constant { rows: Ok(vec![]) });
```
Comment (frankmcsherry) [inline src/dataflow-types/src/plan/mod.rs:441]: I don't understand this. Don't we end up deleting the input and replacing it with the empty collection?

**Review by frankmcsherry — APPROVED** [review]
Looks good module confused questions. Is the test that this fixes already added, from earlier PRs?

File: src/dataflow-types/src/plan/mod.rs (hunk 3 above)
Comment (frankmcsherry) [reply src/dataflow-types/src/plan/mod.rs:441]: Oh, wait. No. I understand how `replace` works now, sorry.

**Comment by umanwizard** [comment]
> Is the test that this fixes already added, from earlier PRs?

Yes, in #9926

File: src/dataflow-types/src/plan/mod.rs (hunk 2 above)
Comment (umanwizard) [reply src/dataflow-types/src/plan/mod.rs:421]: I think that's right, but I don't think rewriting `difference` is worth the effort to save a few clones on a cold path.

Comment (umanwizard) [reply src/dataflow-types/src/plan/mod.rs:421]: I can do it if you find it important, though.

**Comment by umanwizard** [comment]
Sorry for the long delay in publishing this; @frankmcsherry it should be ready for another look.

I've refactored it in the following ways:

1. Permuting maps are always generated at the same point as thinning
   expressions, to ensure that if we refactor the relevant function,
   the logic will not get out of sync. There is a caveat; see below
   for details.

2. We know, without exception, know during LIR planning which "form" of a collection should be used.
   "Form" means either the raw, unarranged collection, or a collection arranged by
   a particular key and requiring a particular permutation to access it.
   I have invented the terminology "collection form" here and am happy to change it
   if a better term is discovered.

3. We now explicitly track the information of whether the "raw" form is available,
   rather than implicitly assuming it is available whenever no arranged keys are available.
   Thus in LIR planning we pass around a an `AvailableCollections` object representing
   a set of forms rather than simply a `Vec` of keys.

4. We now never create arrangements during rendering if they are not explicitly named in LIR.
   Instead, we create them by creating the necessary ArrangeBy node during LIR planning;
   i.e., we do for every operator that relies on an unplanned arrangement
   what was done in https://github.com/MaterializeInc/materialize/pull/10021
   for delta joins.
   The exception is arrangements that can't be named in LIR, currently just the
   arrangement of the `stream` collection for joins.

5. `ArrangeBy` can now turn an arbitrary set of forms into another arbitrary set of forms.
   To put that more concretely: it is now able to form the raw collection from an arranged
   collection, and not just the other way around. Perhaps we could come up with some
   more accurate name for it than `ArrangeBy`.

## Caveats:

1. We don't communicate the permutation/thinning information across different
   acts of building dataflows.
   If we export an index with a given key from one dataflow, we only save the key,
   and then re-form the permutation when we import it into another dataflow.
   This would break if, for example, we come up with some other function that creates
   permutations in some other way, and only use it for some dataflows, but forget to
   account for that at the site where we re-form the permutation. This seems unlikely
   to cause problems in the foreseeable future, but it should still be fixed for reasons of
   code elegance/clarity; doing so will probably require storing the permutations in
   the catalog and DataflowDescription.

2. We are still passing around a column permutation `HashMap`, rather than a `MapFilterProject`
   as I had hoped we could do. The reason we can't do so is because there are a few cases where
   we need to permute something that isn't an MFP. For example, we need to permute each of
   a `FlatMap`'s input expressions.

3. I haven't yet attempted to fix the internal logging dataflows, so they still use
   the otherwise-irrelevant `make_thinning_expression` function,
   but I will do so as a quick follow-up (hopefully before landing this PR)

**Pull request merged**